### DISCUSSION TITLE
Resolving "function" parameters when the output struct has fields of the same name

### DISCUSSION THREAD

**Discussion by weberc2** [cuelang/cue#967] [Q&A]
I'm a newbie (I began learning Cue this afternoon). I have a related question about resolving parameter names when the function parameters have the same names as the output struct (I've scoured the docs and done a bunch of Googling, but it's entirely likely that being a newbie, I'm searching for the wrong criteria):

```cue
// In Python, I would do this:
//
// def Job(command, args): return f"{command} {args}"
// def GoJob(command, args): return Job("go", f"{command} {args}")

// I can use different parameter names for GoJob and Job (e.g., command vs
// command_), but this gets kind of obnoxious after a few levels of nesting
// (e.g., command___ vs command____).

#Job: {
    #command: string
    #args: string
    "\(#command) \(#args)"
}

#GoJob: {
    // These are what I want to reference in the string interpolation below.
    // Instead, I end up referencing the resolved values.
    #command: string
    #args: string

    #Job & {
        #command: "go"
        // I've also tried `#args: "\(#command) \(#args)"` but to no avail
        #args: "\(#GoJob.#command) \(#GoJob.#args)"
    }
}

job: #GoJob & {
    // job.#command: conflicting values "test" and "go"
    // job.#command: invalid interpolation: conflicting values "test" and "go"
    #command: "test"
    #args: "-v ./..."
}
```

**Comment by seh** [comment]
To solve the interpolation problem in `#GoJob`, you can [use a value alias](https://tip.cuelang.org/play/?id=qu7RCEq8CdJ#cue@export@cue). However, that doesn't get past the problem of the `#command` and `#args` fields in `job` seemingly conflicting inappropriately with the synonyms in `#Job`.

Note that if you[ disambiguate the names, it works as intended](https://tip.cuelang.org/play/?id=XkMj5SK8REQ#cue@export@cue), but that doesn't seem like it should be necessary here.

**Reply by myitcv** [reply]
> but that doesn't seem like it should be necessary here

AFAICT, the main problem here is the embedding of `#Job`. That causes the two `#command` fields to conflict (same with the `#args`) when used in `job`. 

I would resolve this as follws:

```cue
#Job: {
    #command: string
    #args: string
    _#Job & {
        _#command: #command
        _#args: #args
    }
}

_#Job: {
    _#command: string
    _#args: string
    "\(_#command) \(_#args)"
}

#GoJob: {
    #command: string
    #args: string

    _#Job & {
        _#command: "go"
        _#args: "\(#command) \(#args)"
        _
    }
}

job: #GoJob & {
    #command: "test"
    #args: "-v ./..."
    _
}
```

https://tip.cuelang.org/play/?id=U2j5KT0gPC3#cue@export@cue

**Reply by seh** [reply]
I see that that works, though by way of introducing another level of embedding, such that `#GoJob` and `#Job` are no longer related. Perhaps that doesn't matter at run time, but it's working around being unable to write what was the natural inclination for the design.

**Reply by myitcv** [reply]
Understood, but I'm not clear how that natural inclination could work in the context of the embedding.

**Reply by weberc2** [reply]
To be clear, embedding isn't a requirement, I just want reuse and I had seen some examples which used embedding. 🙂

Basically I want to know how to write a function in cue?

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/967.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
